### ISSUE TITLE
FakeResolver tsan issue

### ISSUE THREAD

**Issue by karthikravis** [grpc/grpc#19688]
### What version of gRPC and what language are you using?
gRPC head. 
 
### What operating system (Linux, Windows,...) and version?
Linux
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
bazel 
 
### What did you do?
Ran presubmit tests with https://github.com/grpc/grpc/pull/19406
 
### What did you expect to see?

No TSAN fauilures 
 
### What did you see instead?

https://source.cloud.google.com/results/invocations/34d491d7-a335-4047-ad30-007b79774031/artifacts/%2F%2Ftest%2Fcpp%2Fclient:client_channel_stress_test@poller%3Depoll1

```
WARNING: ThreadSanitizer: data race (pid=18)
  Write of size 8 at 0x7b4000020918 by thread T1:
    #0 grpc_core::FakeResolver::FakeResolver(grpc_core::ResolverArgs) /proc/self/cwd/src/core/ext/filters/client_channel/resolver/fake/fake_resolver.cc:102 (liblibgrpc_Uresolver_Ufake.so+0x7981)
    #1 grpc_core::FakeResolver* grpc_core::New<grpc_core::FakeResolver, grpc_core::ResolverArgs>(grpc_core::ResolverArgs&&) /proc/self/cwd/./src/core/lib/gprpp/memory.h:51 (liblibgrpc_Uresolver_Ufake.so+0xca1e)
    #2 grpc_core::(anonymous namespace)::FakeResolverFactory::CreateResolver(grpc_core::ResolverArgs) const /proc/self/cwd/src/core/ext/filters/client_channel/resolver/fake/fake_resolver.cc:326 (liblibgrpc_Uresolver_Ufake.so+0x945f)
    #3 grpc_core::ResolverRegistry::CreateResolver(char const*, grpc_channel_args const*, grpc_pollset_set*, grpc_combiner*, std::unique_ptr<grpc_core::Resolver::ResultHandler, grpc_core::DefaultDelete<grpc_core::Resolver::ResultHandler> >) /proc/self/cwd/src/core/ext/filters/client_channel/resolver_registry.cc:163 (liblibgrpc_Uclient_Uchannel.so+0xa57ca)
    #4 grpc_core::ResolvingLoadBalancingPolicy::ResolvingLoadBalancingPolicy(grpc_core::LoadBalancingPolicy::Args, grpc_core::TraceFlag*, std::unique_ptr<char, grpc_core::DefaultDelete<char> >, bool (*)(void*, grpc_core::Resolver::Result const&, char const**, grpc_core::RefCountedPtr<grpc_core::LoadBalancingPolicy::Config>*, grpc_error**), void*) /proc/self/cwd/src/core/ext/filters/client_channel/resolving_lb_policy.cc:194 (liblibgrpc_Uclient_Uchannel.so+0xb08ec)
    #5 grpc_core::ResolvingLoadBalancingPolicy* grpc_core::New<grpc_core::ResolvingLoadBalancingPolicy, grpc_core::LoadBalancingPolicy::Args, grpc_core::TraceFlag*, std::unique_ptr<char, grpc_core::DefaultDelete<char> >, bool (&)(void*, grpc_core::Resolver::Result const&, char const**, grpc_core::RefCountedPtr<grpc_core::LoadBalancingPolicy::Config>*, grpc_error**), grpc_core::(anonymous namespace)::ChannelData*>(grpc_core::LoadBalancingPolicy::Args&&, grpc_core::TraceFlag*&&, std::unique_ptr<char, grpc_core::DefaultDelete<char> >&&, bool (&)(void*, grpc_core::Resolver::Result const&, char const**, grpc_core::RefCountedPtr<grpc_core::LoadBalancingPolicy::Config>*, grpc_error**), grpc_core::(anonymous namespace)::ChannelData*&&) /proc/self/cwd/./src/core/lib/gprpp/memory.h:51 (liblibgrpc_Uclient_Uchannel.so+0x753f5)
    #6 grpc_core::(anonymous namespace)::ChannelData::CreateResolvingLoadBalancingPolicyLocked() /proc/self/cwd/src/core/ext/filters/client_channel/client_channel.cc:1273 (liblibgrpc_Uclient_Uchannel.so+0x750a5)
    #7 grpc_core::(anonymous namespace)::ChannelData::TryToConnectLocked(void*, grpc_error*) /proc/self/cwd/src/core/ext/filters/client_channel/client_channel.cc:1583 (liblibgrpc_Uclient_Uchannel.so+0x74eb5)
    #8 closure_impl::closure_wrapper(void*, grpc_error*) /proc/self/cwd/./src/core/lib/iomgr/closure.h:151 (liblibgrpc++.so+0x2acad)
    #9 grpc_combiner_continue_exec_ctx() /proc/self/cwd/src/core/lib/iomgr/combiner.cc:268 (liblibgrpc_Ubase_Uc.so+0x8a92c)
    #10 grpc_core::ExecCtx::Flush() /proc/self/cwd/src/core/lib/iomgr/exec_ctx.cc:151 (liblibgrpc_Ubase_Uc.so+0xb1a2a)
    #11 grpc_core::Executor::RunClosures(char const*, grpc_closure_list) /proc/self/cwd/src/core/lib/iomgr/executor.cc:145 (liblibgrpc_Ubase_Uc.so+0xb302b)
    #12 grpc_core::Executor::ThreadMain(void*) /proc/self/cwd/src/core/lib/iomgr/executor.cc:265 (liblibgrpc_Ubase_Uc.so+0xb38df)
    #13 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::{lambda(void*)#1}::operator()(void*) const /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:140 (liblibgpr_Ubase.so+0x16b54)
    #14 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::{lambda(void*)#1}::__invoke(void*) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:110 (liblibgpr_Ubase.so+0x16968)

  Previous read of size 8 at 0x7b4000020918 by main thread:
    #0 grpc_core::FakeResolverResponseGenerator::SetResponse(grpc_core::Resolver::Result) /proc/self/cwd/src/core/ext/filters/client_channel/resolver/fake/fake_resolver.cc:188 (liblibgrpc_Uresolver_Ufake.so+0x7cf0)
    #1 grpc_core::(anonymous namespace)::GrpcLb::ProcessAddressesAndChannelArgsLocked(grpc_core::InlinedVector<grpc_core::ServerAddress, 1ul> const&, grpc_channel_args const&) /proc/self/cwd/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc:1460 (liblibgrpc_Ulb_Upolicy_Ugrpclb_Usecure.so+0x24811)
    #2 grpc_core::(anonymous namespace)::GrpcLb::UpdateLocked(grpc_core::LoadBalancingPolicy::UpdateArgs) /proc/self/cwd/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc:1380 (liblibgrpc_Ulb_Upolicy_Ugrpclb_Usecure.so+0x1ed4d)
    #3 grpc_core::ResolvingLoadBalancingPolicy::CreateOrUpdateLbPolicyLocked(char const*, grpc_core::RefCountedPtr<grpc_core::LoadBalancingPolicy::Config>, grpc_core::Resolver::Result, grpc_core::InlinedVector<char*, 3ul>*) /proc/self/cwd/src/core/ext/filters/client_channel/resolving_lb_policy.cc:371 (liblibgrpc_Uclient_Uchannel.so+0xb1ca9)
    #4 grpc_core::ResolvingLoadBalancingPolicy::OnResolverResultChangedLocked(grpc_core::Resolver::Result) /proc/self/cwd/src/core/ext/filters/client_channel/resolving_lb_policy.cc:486 (liblibgrpc_Uclient_Uchannel.so+0xb2a49)
    #5 grpc_core::ResolvingLoadBalancingPolicy::ResolverResultHandler::ReturnResult(grpc_core::Resolver::Result) /proc/self/cwd/src/core/ext/filters/client_channel/resolving_lb_policy.cc:87 (liblibgrpc_Uclient_Uchannel.so+0xb464c)
    #6 grpc_core::FakeResolver::MaybeSendResultLocked() /proc/self/cwd/src/core/ext/filters/client_channel/resolver/fake/fake_resolver.cc:152 (liblibgrpc_Uresolver_Ufake.so+0x83a3)
    #7 grpc_core::FakeResolverResponseGenerator::SetResponseLocked(void*, grpc_error*) /proc/self/cwd/src/core/ext/filters/client_channel/resolver/fake/fake_resolver.cc:182 (liblibgrpc_Uresolver_Ufake.so+0x8725)
    #8 grpc_combiner_continue_exec_ctx() /proc/self/cwd/src/core/lib/iomgr/combiner.cc:268 (liblibgrpc_Ubase_Uc.so+0x8a92c)
    #9 grpc_core::ExecCtx::Flush() /proc/self/cwd/src/core/lib/iomgr/exec_ctx.cc:151 (liblibgrpc_Ubase_Uc.so+0xb1a2a)
    #10 grpc_core::ExecCtx::~ExecCtx() /proc/self/cwd/./src/core/lib/iomgr/exec_ctx.h:121 (client_channel_stress_test+0x4ff69c)
    #11 grpc::testing::(anonymous namespace)::ClientChannelStressTest::SetNextResolution(std::vector<grpc::testing::(anonymous namespace)::ClientChannelStressTest::AddressData, std::allocator<grpc::testing::(anonymous namespace)::ClientChannelStressTest::AddressData> > const&) /proc/self/cwd/test/cpp/client/client_channel_stress_test.cc:245 (client_channel_stress_test+0x4d51fc)
    #12 grpc::testing::(anonymous namespace)::ClientChannelStressTest::Run() /proc/self/cwd/test/cpp/client/client_channel_stress_test.cc:159 (client_channel_stress_test+0x4d273c)
    #13 main /proc/self/cwd/test/cpp/client/client_channel_stress_test.cc:335 (client_channel_stress_test+0x4d211a)

  Location is heap block of size 256 at 0x7b4000020900 allocated by thread T1:
    #0 malloc /tmp/clang-build/src/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:672 (client_channel_stress_test+0x462054)
    #1 gpr_malloc /proc/self/cwd/src/core/lib/gpr/alloc.cc:57 (liblibgpr_Ubase.so+0x8e25)
    #2 grpc_core::FakeResolverResponseGenerator* grpc_core::New<grpc_core::FakeResolverResponseGenerator>() /proc/self/cwd/./src/core/lib/gprpp/memory.h:50 (client_channel_stress_test+0x4faf3a)
    #3 grpc_core::RefCountedPtr<grpc_core::FakeResolverResponseGenerator> grpc_core::MakeRefCounted<grpc_core::FakeResolverResponseGenerator>() /proc/self/cwd/./src/core/lib/gprpp/ref_counted_ptr.h:180 (client_channel_stress_test+0x4fabb3)
    #4 grpc_core::(anonymous namespace)::GrpcLb::GrpcLb(grpc_core::LoadBalancingPolicy::Args) /proc/self/cwd/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc:1283 (liblibgrpc_Ulb_Upolicy_Ugrpclb_Usecure.so+0x1d9bb)
    #5 grpc_core::(anonymous namespace)::GrpcLb* grpc_core::New<grpc_core::(anonymous namespace)::GrpcLb, grpc_core::LoadBalancingPolicy::Args>(grpc_core::LoadBalancingPolicy::Args&&) /proc/self/cwd/./src/core/lib/gprpp/memory.h:51 (liblibgrpc_Ulb_Upolicy_Ugrpclb_Usecure.so+0x1d81e)
    #6 grpc_core::(anonymous namespace)::GrpcLbFactory::CreateLoadBalancingPolicy(grpc_core::LoadBalancingPolicy::Args) const /proc/self/cwd/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc:1771 (liblibgrpc_Ulb_Upolicy_Ugrpclb_Usecure.so+0x1d10f)
    #7 grpc_core::LoadBalancingPolicyRegistry::CreateLoadBalancingPolicy(char const*, grpc_core::LoadBalancingPolicy::Args) /proc/self/cwd/src/core/ext/filters/client_channel/lb_policy_registry.cc:94 (liblibgrpc_Uclient_Uchannel.so+0x9e099)
    #8 grpc_core::ResolvingLoadBalancingPolicy::CreateLbPolicyLocked(char const*, grpc_channel_args const&, grpc_core::InlinedVector<char*, 3ul>*) /proc/self/cwd/src/core/ext/filters/client_channel/resolving_lb_policy.cc:387 (liblibgrpc_Uclient_Uchannel.so+0xb1efa)
    #9 grpc_core::ResolvingLoadBalancingPolicy::CreateOrUpdateLbPolicyLocked(char const*, grpc_core::RefCountedPtr<grpc_core::LoadBalancingPolicy::Config>, grpc_core::Resolver::Result, grpc_core::InlinedVector<char*, 3ul>*) /proc/self/cwd/src/core/ext/filters/client_channel/resolving_lb_policy.cc:349 (liblibgrpc_Uclient_Uchannel.so+0xb19b8)
    #10 grpc_core::ResolvingLoadBalancingPolicy::OnResolverResultChangedLocked(grpc_core::Resolver::Result) /proc/self/cwd/src/core/ext/filters/client_channel/resolving_lb_policy.cc:486 (liblibgrpc_Uclient_Uchannel.so+0xb2a49)
    #11 grpc_core::ResolvingLoadBalancingPolicy::ResolverResultHandler::ReturnResult(grpc_core::Resolver::Result) /proc/self/cwd/src/core/ext/filters/client_channel/resolving_lb_policy.cc:87 (liblibgrpc_Uclient_Uchannel.so+0xb464c)
    #12 grpc_core::FakeResolver::MaybeSendResultLocked() /proc/self/cwd/src/core/ext/filters/client_channel/resolver/fake/fake_resolver.cc:152 (liblibgrpc_Uresolver_Ufake.so+0x83a3)
    #13 grpc_core::FakeResolverResponseGenerator::SetResponseLocked(void*, grpc_error*) /proc/self/cwd/src/core/ext/filters/client_channel/resolver/fake/fake_resolver.cc:182 (liblibgrpc_Uresolver_Ufake.so+0x8725)
    #14 grpc_combiner_continue_exec_ctx() /proc/self/cwd/src/core/lib/iomgr/combiner.cc:268 (liblibgrpc_Ubase_Uc.so+0x8a92c)
    #15 grpc_core::ExecCtx::Flush() /proc/self/cwd/src/core/lib/iomgr/exec_ctx.cc:151 (liblibgrpc_Ubase_Uc.so+0xb1a2a)
    #16 grpc_core::Executor::RunClosures(char const*, grpc_closure_list) /proc/self/cwd/src/core/lib/iomgr/executor.cc:145 (liblibgrpc_Ubase_Uc.so+0xb302b)
    #17 grpc_core::Executor::ThreadMain(void*) /proc/self/cwd/src/core/lib/iomgr/executor.cc:265 (liblibgrpc_Ubase_Uc.so+0xb38df)
    #18 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::{lambda(void*)#1}::operator()(void*) const /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:140 (liblibgpr_Ubase.so+0x16b54)
    #19 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::{lambda(void*)#1}::__invoke(void*) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:110 (liblibgpr_Ubase.so+0x16968)

  Thread T1 'default-executo' (tid=20, running) created by main thread at:
    #0 pthread_create /tmp/clang-build/src/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:976 (client_channel_stress_test+0x4637eb)
    #1 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:109 (liblibgpr_Ubase.so+0x164d2)
    #2 grpc_core::(anonymous namespace)::ThreadInternalsPosix* grpc_core::New<grpc_core::(anonymous namespace)::ThreadInternalsPosix, char const*&, void (*&)(void*), void*&, bool*, grpc_core::Thread::Options const&>(char const*&, void (*&)(void*), void*&, bool*&&, grpc_core::Thread::Options const&) /proc/self/cwd/./src/core/lib/gprpp/memory.h:51 (liblibgpr_Ubase.so+0x15e78)
    #3 grpc_core::Thread::Thread(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:186 (liblibgpr_Ubase.so+0x15bef)
    #4 grpc_core::Executor::SetThreading(bool) /proc/self/cwd/src/core/lib/iomgr/executor.cc:181 (liblibgrpc_Ubase_Uc.so+0xb2637)
    #5 grpc_core::Executor::Init() /proc/self/cwd/src/core/lib/iomgr/executor.cc:113 (liblibgrpc_Ubase_Uc.so+0xb205d)
    #6 grpc_core::Executor::InitAll() /proc/self/cwd/src/core/lib/iomgr/executor.cc:415 (liblibgrpc_Ubase_Uc.so+0xb47c8)
    #7 grpc_iomgr_init() /proc/self/cwd/src/core/lib/iomgr/iomgr.cc:60 (liblibgrpc_Ubase_Uc.so+0xba7c8)
    #8 grpc_init /proc/self/cwd/src/core/lib/surface/init.cc:144 (liblibgrpc.so+0x3d98)
    #9 main /proc/self/cwd/test/cpp/client/client_channel_stress_test.cc:332 (client_channel_stress_test+0x4d20e3)

SUMMARY: ThreadSanitizer: data race /proc/self/cwd/src/core/ext/filters/client_channel/resolver/fake/fake_resolver.cc:102 in grpc_core::FakeResolver::FakeResolver(grpc_core::ResolverArgs)
 ```

**Comment by markdroth** [comment]
QC, could this have been caused by your recent change to the fake resolver?

**Comment by bigfacebear** [comment]
I think this is not caused by my recent change since it has not been merged into master.
But I believe this is caused by the problem what I am fixing now.
`Write of size 8 at 0x7b4000020918 by thread T1` should be setting `FakeResolverResponseGenerator::resolver_` in the ctor of `FakeResolver`. `Previous read of size 8 at 0x7b4000020918 by main thread` should be checking whether `FakeResolverResponseGenerator::resolver_` is nullptr in `FakeResolverResponseGenerator::SetResponse()`.
I think this can be solved after #19634 being merged. Or I can open a separate PR for this?

**Comment by markdroth** [comment]
If you think #19634 can be merged quickly, then it's fine to wait for that.  If you think it will take a while to get done, then it might be a good idea to pull out this change into a separate PR.

Thanks!

**Comment by bigfacebear** [comment]
This issue should have been solved since #19634 is merged.
